Added dataset list for usecase

adp-atea opened the pull request.

adp-atea committed:
Added dataset list for usecase

adp-atea committed:
Tests, some fixes

adp-atea committed:
Fixed bug where .add() would remove all other datasets from the list

sirex commented on vitrina/projects/templates/vitrina/projects/datasets.html:
```diff
@@ -0,0 +1,85 @@
+                        <div class="content">
+                            <div>
+                                <strong>
+                                    <a href="{{ i.object.get_absolute_url }}"class="dataset-list-title">
```
This link does not work, `i.object` does not exist.

sirex commented on vitrina/projects/views.py:
```diff
@@ -93,4 +104,71 @@ class ProjectHistoryView(HistoryView):
+        return context_data
+
+
+class RemoveDatasetView(LoginRequiredMixin, PermissionRequiredMixin, View):
```
Maybe use `DeleteView` with `confirm_delete.html` template as in `DeleteMemberView` for better consistency?

sirex commented on vitrina/projects/views.py:
```diff
@@ -93,4 +104,71 @@ class ProjectHistoryView(HistoryView):
+
+class RemoveDatasetView(LoginRequiredMixin, PermissionRequiredMixin, View):
+    def has_permission(self):
+        project = get_object_or_404(Project, pk=self.kwargs.get('project_id'))
```
Do not query same thing multiple times, use `dispatch` as in `ProjectDatasetsView`.

sirex commented on vitrina/projects/views.py:
```diff
@@ -93,4 +104,71 @@ class ProjectHistoryView(HistoryView):
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_queryset(self):
+        return self.object.datasets.all()
```
Use public manager to filter out deleted ant non-public datasets:

https://docs.djangoproject.com/en/3.2/topics/db/queries/#using-a-custom-reverse-manager

sirex commented on vitrina/projects/forms.py:
```diff
@@ -12,6 +13,11 @@ class ProjectForm(ModelForm):
+        label=_('Duomenų rinkiniai'),
+        queryset=Dataset.objects.all(),
+        required=False
+    )
```
Do not show this field, because there are too many datasets to select them this way, via a field. Datasets and projects should be associated from dataset details page, see https://github.com/atviriduomenys/katalogas/issues/239.

sirex commented on vitrina/projects/templates/vitrina/projects/datasets.html:
```diff
@@ -0,0 +1,85 @@
+                                    </a>
+                                </strong>
+                            </div>
+                            {% if dataset.organization_title %}
```
There is not such thing as `organization_title`. Probably you mean `organization.title`? Also if query takes data from other related models, don't forget to add select_related:

https://docs.djangoproject.com/en/3.2/ref/models/querysets/#select-related

sirex commented on vitrina/projects/urls.py:
```diff
@@ -10,6 +10,10 @@
+    path('projects/<int:pk>/datasets/', ProjectDatasetsView.as_view(), name='project-datasets'),
+    path('projects/<int:project_id>/datasets/<int:dataset_id>/remove/', RemoveDatasetView.as_view(),
+         name='project-dataset-remove'),
+    path('projects/<int:pk>/add/', ProjectAddDatasetView.as_view(), name='project-dataset-add'),
```
```suggestion
    path('projects/<int:pk>/datasets/add/', ProjectAddDatasetView.as_view(), name='project-dataset-add'),
```

sirex commented on vitrina/projects/forms.py:
```diff
@@ -26,7 +32,29 @@ def __init__(self, *args, **kwargs):
         )
+
+
+class AddDatasetForm(ModelForm):
```
We don't need this form and possibility to add datasets from project view, because there are too many datasets and this way of adding datasets to projects is not practical, nor convenient.

Instead, projects and datasets should be associated from datasets details view, see https://github.com/atviriduomenys/katalogas/issues/239.

adp-atea committed:
Merge branch 'devel' into 191-duomenų-rinkinių-sąrašas-projekte

adp-atea committed:
Added project list in dataset details page, changed so you choose projects for datasets and not datasets for projects

adp-atea committed:
Merge branch 'devel' into 191-duomenų-rinkinių-sąrašas-projekte

# Conflicts:
#	tests/projects/test_views.py
#	vitrina/projects/models.py
#	vitrina/projects/templates/vitrina/projects/detail.html
#	vitrina/projects/views.py

adp-atea committed:
Made it so authot sees only his projects in add project view and staff or superusers see all

adp-atea committed:
Added tests, fixed bugs

adp-atea committed:
Fixed remove bug, added remove tests

adp-atea committed:
added if statement for removal of project by creators of the project

sirex committed:
Show project titles in dataset details view

sirex committed:
Exclude rejected projects

sirex committed:
Properly handle unauthorized users in dataset projects

sirex committed:
Multiple fixes for dataset project list

- Fixed comment message, when a project is added to the dataset.

- Tried to filter out projects that are already added to the dataset in
  "Add project" form.

- Added a new permission ADD_PROJECT, for adding new projects. UPDATE
  permission is too broad.

- Fixed minor template tings, like added parent title, a message when
  dataset does not have assigned projects, etc.

sirex commented on vitrina/datasets/forms.py:
```diff
@@ -101,6 +102,47 @@ def __init__(self, *args, **kwargs):
+                # FIXME: This does not work, somehow projects, that are
+                #        already assigned to the dataset should be excluded
+                #        from list of options for this field.
+                datasets=self.dataset,
```
Please fix this, in the dataset project form, should be shown only projects, that are not already added to the dataset. I tried to fix it, but a quick attempt didn't worked, so could you please fix it for me? Also do the same thing when checking if use has projects or not, I will mention it in another comment.

sirex commented on vitrina/datasets/views.py:
```diff
@@ -491,3 +493,118 @@ def get_success_url(self):
+    def get_queryset(self):
+        return (
+            Project.public.
+            filter(datasets=self.object).
```
Exclude rejected projects.

sirex commented on vitrina/datasets/views.py:
```diff
@@ -491,3 +493,118 @@ def get_success_url(self):
+        if self.request.user.is_authenticated:
+            context['has_projects'] = (
+                Project.public.
+                filter(user=self.request.user).
```
Exclude rejected and projects that are already added to this dataset. Probably this should be moved to a service function, because it is used in more than one place.

sirex commented on vitrina/datasets/views.py:
```diff
@@ -491,3 +493,118 @@ def get_success_url(self):
+        return super().dispatch(request, *args, **kwargs)
+
+    def has_permission(self):
+        return has_perm(self.request.user, Action.DELETE, self.project)
```
This is not a project delete action, but rather project add action or rather managing project. So action should be UPDATE.

sirex commented on vitrina/orgs/services.py:
```diff
@@ -22,13 +22,14 @@ class Action(Enum):
     VIEW = 'view'
     HISTORY_VIEW = 'history_view'
     COMMENT = "comment_with_status"
+    ADD_PROJECT = "add_project"
```
I added this new permission, but then realized, that a new permission should not be added, but for all dataset project managment tasks should be used `has_perm(user, Action.UPDATE, project)` check. Because essentially this is a project modification action. So please remove this permission and replace all places to use `has_perm(user, Action.UPDATE, project)`. Make sue that third argument to `has_perm` is correct.

sirex commented on vitrina/orgs/services.py:
```diff
@@ -41,6 +42,7 @@ class Role(Enum):
     (Dataset, Action.UPDATE): [Role.COORDINATOR, Role.MANAGER],
     (Dataset, Action.DELETE): [Role.COORDINATOR, Role.MANAGER],
     (Dataset, Action.HISTORY_VIEW): [Role.COORDINATOR, Role.MANAGER],
+    (Dataset, Action.ADD_PROJECT): [Role.ALL],
```
This should be removed as well.

sirex commented on vitrina/projects/views.py:
```diff
@@ -111,4 +117,51 @@ class ProjectHistoryView(HistoryView):
+    template_name = 'vitrina/projects/datasets.html'
+    paginate_by = 20
+
+    object: Dataset
```
```suggestion
    object: Project
```

But I would probably not use `object` name here, because this is a `MultipleObject` mixin, and `object` is from `SingleObject` mixin, so using both `model` and `object` here is very confusing.

Maybe rename `object` → `project`? To avoid confusion?

adp-atea committed:
Added service function for fetching projects

adp-atea committed:
Merge branch 'devel' into 191-duomenų-rinkinių-sąrašas-projekte

# Conflicts:
#	tests/datasets/test_views.py
#	vitrina/datasets/views.py
#	vitrina/projects/models.py

sirex commented on vitrina/projects/templates/vitrina/projects/datasets.html:
```diff
@@ -0,0 +1,76 @@
+
+{% block current_title %}
+{% translate "Duomenų rinkinių sąrašas" %}
+{% endblock %}
```
Also add `parent_title` which should be dataset title.

adp-atea committed:
Merge branch 'devel' into 191-duomenų-rinkinių-sąrašas-projekte

# Conflicts:
#	vitrina/datasets/forms.py
#	vitrina/projects/forms.py
#	vitrina/projects/models.py

adp-atea committed:
Added parent title for dataset list as project.title or project.url if title is not set

adp-atea committed:
Added parent title to project list under dataset
